File: 8-string-to-integer-atoi/string-to-integer-atoi.py
class Solution(object):
    def myAtoi(self, s):
        i = 0
        sign = 1
        result = 0

        # Skip spaces
        while i < len(s) and s[i] == ' ':
            i += 1

        # Check sign
        if i < len(s) and s[i] == '-':
            sign = -1
            i += 1
        elif i < len(s) and s[i] == '+':
            i += 1

        # Read digits
        while i < len(s) and s[i].isdigit():
            result = result * 10 + int(s[i])
            i += 1

        result *= sign

        # 32-bit range
        if result > 2147483647:
            return 2147483647

        if result < -2147483648:
            return -2147483648

        return result

File: 8-string-to-integer-atoi/string-to-integer-atoi.c
#include <limits.h>

int myAtoi(char* s) {
    int i = 0;
    int sign = 1;
    int result = 0;

    // Skip spaces
    while (s[i] == ' ')
        i++;

    // Check sign
    if (s[i] == '-') {
        sign = -1;
        i++;
    } else if (s[i] == '+') {
        i++;
    }

    // Read digits
    while (s[i] >= '0' && s[i] <= '9') {
        int digit = s[i] - '0';

        // Check overflow
        if (result > INT_MAX / 10 ||
            (result == INT_MAX / 10 && digit > 7)) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }

        result = result * 10 + digit;
        i++;
    }

    return result * sign;
}